### PR TITLE
Create a developer's readme

### PR DESCRIPTION
**Pull request by aesteve-rh**
Create a new developer's README similar to other projects (e.g., ovirt-imageio). This change slightly improves readability
information's organization. In the update is important to keep the files linked so that it is easy to find information
and navigate between the different documentation files.

This PR includes an update of the tests README to markdown format and the update of some sections before
moving them to the new developer's document. Here's a summary of the changes:
1. Transform the **tests** `README` file to `README.md` and change format to fully comply with Markdown syntax.
2. Remove outdated parts from the original README
3. Create a developer's readme in the doc folder and move all developer-specific sections from main `README.md` and `test/README.md` to `doc/development.md`, and linking the new document from main README
4. Improve new development readme:
    - Add `copr virt-preview` enablement to the developer's readme
    - Update storage README "setting up" section to use `make storage` instead of direct script invocation
5. Create and document a new 'clean-storage' make target

Related: #42

### CODE REVIEW

File: doc/development.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+are not covered currently by Jenkins CI. This can be achieved in a fast
+way manually and independently from gerrit by invoking Travis CI on your
+github branch:
+
+- Fork the project on github.
+- Visit https://travis-ci.org, register using your github account, and
+  enable builds for your Vdsm fork.
+- Push your changes to your github fork to trigger a build.
+
+See .travis.yml file for tested Travis platforms and tests configurations.
```
Comment (nirs) [inline doc/development.md:107]: We don't use Travis CI for a while, so we can drop this section. This is best done in a seaprate patch, before creating doc/developement.md

File: doc/development.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+Create storage for the tests:
+
+    make storage
+
+Alternatively, storage tests can be completely setup and run by doing:
+
+    make tests-storage
+
+See the tests [README](../tests/README.md#running-the-tests)
+for more information.
```
Comment (nirs) [inline doc/development.md:54]: Adding the section about creating storage for tests and running storage tests is not related to moving the text from the main readme. It is best done in a separate commit.

File: README.md [hunk 3]
```diff
@@ -63,41 +63,6 @@ To inspect Vdsm service status:
...
-    deactivate
-
-Before running vdsm tests, activate the environment:
-
-    source ~/.venv/vdsm/bin/activate
-
-When done, you can deactivate the environment:
-
-    deactivate
-
```
Comment (nirs) [inline README.md:100]: Removing this to special readme is good, but the new shiny readme should be linked from this file. Maybe add very short section on "Contributing" where we point the read to the new readme for developers.

File: doc/development.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+
+## CI
+
+Running tests locally is convenient, but before your changes can be
+merged, we need to test them on all supported distributions and
+architectures.
+
+When you submit patches to gerrit, oVirt's Jenkins CI will run its tests
+according to configuration in the stdci.yaml file.
```
Comment (nirs) [inline doc/development.md:92]: This section is outdated, but it was before your moved the text, so it is fine to move the text here as is.

File: tests/storage/README.md [hunk 5]
```diff
@@ -9,18 +9,29 @@ or fail (depending on the test).
...
+
+Some storage may not be available on all environments. For example on
+CentOS 7 userstorage cannot create storage with 4k sector size, and in
+oVirt CI, this usually fails. Tests should check if storage exists and
+mark test as xfail or skip when the storage is not available.
+
 If you want to tear down the user storage, run:
 
-    python tests/storage/userstorage.py teardown
+    make clean-storage
```
Comment (nirs) [inline tests/storage/README.md:28]: We don't have yet clean-storage target, but it can be good idea to add one.

File: tests/storage/README.md [hunk 6]
```diff
@@ -9,18 +9,29 @@ or fail (depending on the test).
...
 
 There is no need to tear down the storage normally. The loop devices are
 backed up by sparse files and do not consume much resources.
 
+For more info on userstorage see
+[userstorage](https://github.com/nirs/userstorage).
```
Comment (nirs) [inline tests/storage/README.md:34]: This is awesome, but we don't use userstorage yet! We use an older version of userstorage.

Can you file a github issue for replacing vdsm version of userstorage with userstorage from pip, like imageio?

File: tests/storage/README.md [hunk 7]
```diff
@@ -9,18 +9,29 @@ or fail (depending on the test).
...
 be run again after rebooting the host.
 
+The storage is configured in the file `storage.py` in the root of the
+project. Tests that use user storage load this configuration file and
+access the storage via the BACKENDS dict.
+
+Some storage may not be available on all environments. For example on
+CentOS 7 userstorage cannot create storage with 4k sector size, and in
+oVirt CI, this usually fails. Tests should check if storage exists and
+mark test as xfail or skip when the storage is not available.
```
Comment (nirs) [inline tests/storage/README.md:24]: This section is outdated (limit of old ovirt ci, based on centos 7) and can be removed.

File: tests/storage/README.md [hunk 8]
```diff
@@ -9,18 +9,29 @@ or fail (depending on the test).
...
 
-    python tests/storage/userstorage.py setup
+    make storage
 
 This can be run once when setting up development environment, and must
 be run again after rebooting the host.
 
+The storage is configured in the file `storage.py` in the root of the
+project. Tests that use user storage load this configuration file and
+access the storage via the BACKENDS dict.
```
Comment (nirs) [inline tests/storage/README.md:19]: This is not true (yet) - this describes how upstream userstorage works (used in imageio) but we still use the old userstorage version (tests/storage/userstorage.py).

**Review by nirs — CHANGES_REQUESTED** [review]
I like this approach, but we need need to make the changes in a way that it is easier to review and improve.

1. Remove outdated parts from the original readme
2. Created the new document, moving code from main readme without any change, and linking to the new document from main readme.
3. Improve new development readme.

File: tests/storage/README.md (hunk 6 above)
Comment (aesteve-rh) [reply tests/storage/README.md]: Done: https://github.com/oVirt/vdsm/issues/151

I'll re-add this link when the issue gets addressed.

File: tests/README.md [hunk 9]
```diff
@@ -13,6 +13,11 @@ Note: Not all tests have been converted to tox and/or pytest yet.
 
 ## Running the tests
 
+Before running storage tests, make sure you have setup the user storage
+accordingly. Read
+[Setting up user storage](./storage/README.md#setting-up-user-storage)
+for more information.
```
Comment (nirs) [inline tests/README.md:19]: It is better to show how to setup the storage here:

    make storage

And link the storage readme for more info.

**Review by nirs — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -154,24 +154,8 @@ Running tests locally is convenient, but before your changes can be
...
-- Fork the project on github.
-- Visit https://travis-ci.org, register using your github account, and
-  enable builds for your Vdsm fork.
-- Push your changes to your github fork to trigger a build.
-
-See .travis.yml file for tested Travis platforms and tests configurations.
-
+When you push patches to GitHub, CI will run its tests according to the
+configuration in the .github/workflows/ci.yml file.
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply README.md]: Uhm, first to clarify: I think I misread, you said it would be better in a different patch (before this one), and I just thought you meant a different commit previous to the split. So sorry about that, I can still revert it and split it if you think it is better that way.

And about removing the section: you wrote that Travis CI was not used for a while, and that we could drop the section. I wanted not just to drop it, but try to update it, and found the GitHub CI config file.

I have documented the change in the commit message, but again, I could drop the change and do it in a separate patch as you initially pointed.

[the remark this answers is not in the document]
Comment (nirs) [reply README.md]: I commented about the missing commit message. I know why you removed this but the person reading git log in 6 month will not have any idea if you don't explain this in the commit message.

**Comment by nirs** [comment]
@aesteve-rh the PR message is *not* a replacement for the commit message for every commit. The PR message is good place to explain how the commits are related, and adding information that can help reviewers and does not belong in the commit message, the code, or documentation.

The text of the commit message is part of git history and is extremely important to understand the changes in the future.

File: doc/development.md [hunk 11]
```diff
@@ -14,6 +14,12 @@ Enable oVirt packages for Fedora:
     sudo dnf copr enable -y nsoffer/ioprocess-preview
     sudo dnf copr enable -y nsoffer/ovirt-imageio-preview
 
+Enable
+[virt-preview](https://copr.fedorainfracloud.org/coprs/g/virtmaint-sig/virt-preview/)
+package to obtain latest qemu and libvirt versions (and update if needed):
```
Comment (nirs) [inline doc/development.md:19]: Enable [virt-preview](...) repository ...

What do you mean by "(and update if needed)"?

File: doc/development.md [hunk 12]
```diff
@@ -26,14 +32,8 @@ Create virtual environment for vdsm:
...
-Before running vdsm tests, activate the environment:
-
-    source ~/.venv/vdsm/bin/activate
-
-When done, you can deactivate the environment:
-
-    deactivate
-
+Read [Vdsm tests](../tests/README.md) for more information on
+how to run tests.
```
Comment (nirs) [inline doc/development.md:36]: I don't think we need this sentence, the next topic is running the tests, and this is a pretty small document.

File: tests/README.md [hunk 13]
```diff
@@ -13,6 +13,23 @@ Note: Not all tests have been converted to tox and/or pytest yet.
...
+    deactivate
+
+Before running storage tests, make sure you have setup the user storage
+accordingly. To create storage you can do:
+
+    make storage
+
+Read
+[Setting up user storage](./storage/README.md#setting-up-user-storage)
+for more information.
```
Comment (nirs) [inline tests/README.md:31]: Read ... for more info is correct, but I think "See ... for more info" would be better.

File: doc/development.md (hunk 11 above)
Comment (aesteve-rh) [reply doc/development.md:19]: Well, after adding the **repository** (I'll fix it!), one may need to do `dnf update`, I doubted if it was worth mentioning it explicitly, or if I could just assume people would know they have to update afterwards, so I ended up somewhere in between.

If it is not clear, I will re-phrase it or explicitly write the command.

Comment (nirs) [reply doc/development.md]: If we need to do "dnf update", it should be a documented step. But I don't remember that this is neededl.

**Review by nirs — CHANGES_REQUESTED** [review]

File: doc/development.md [hunk 14]
```diff
@@ -14,6 +14,13 @@ Enable oVirt packages for Fedora:
     sudo dnf copr enable -y nsoffer/ioprocess-preview
     sudo dnf copr enable -y nsoffer/ovirt-imageio-preview
 
+Enable
+[virt-preview](https://copr.fedorainfracloud.org/coprs/g/virtmaint-sig/virt-preview/)
+package to obtain latest qemu and libvirt versions:
```
[the remark this answers is not in the document]
Comment (aesteve-rh) [reply doc/development.md:19]: :facepalm: I was paying attention to do the changes to the right commit and forgot to change this. My bad.

File: doc/development.md (hunk 11 above)
Comment (aesteve-rh) [reply doc/development.md]: Well, adding the repository alone does not update qemu and libvirt versions if it is required. I had to run `sudo dnf update` myself if I remember it correctly.

Comment (nirs) [reply doc/development.md]: But these steps are done before installing vdsm - when you install vdsm it will pull the right version of qemu and libvirt from the virt preview repo.

Testing this require starting with a clean host and running the steps one by one.

File: README.md [hunk 15]
```diff
@@ -64,40 +69,6 @@ To inspect Vdsm service status:
...
-
-Before running vdsm tests, activate the environment:
-
-    source ~/.venv/vdsm/bin/activate
-
-When done, you can deactivate the environment:
-
-    deactivate
-
 ## Building Vdsm
```
Comment (nirs) [inline README.md:72]: Building vdsm from git belongs to the developer readme. Users are not expected to install vdsm from git.

**Review by nirs — CHANGES_REQUESTED** [review]
Current changes looks good, but more content need to move the development readme.

**Review by mz-pdm — APPROVED** [review]
Thanks for the updates!

(More updates may be needed, but it's difficult to work with large PRs, so it's better to merge what we have and leave further stuff for followup patches.)

**Review by nirs — APPROVED** [review]
